compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/PushBackDoFnTransform.java

```diff
@@ -0,0 +1,160 @@
+
+  /**
+   * PushBackDoFnTransform Constructor.
+   */
+  public PushBackDoFnTransform(final DoFn<InputT, OutputT> doFn,
+                               final Coder<InputT> inputCoder,
+                               final Map<TupleTag<?>, Coder<?>> outputCoders,
+                               final TupleTag<OutputT> mainOutputTag,
+                               final List<TupleTag<?>> additionalOutputTags,
+                               final WindowingStrategy<?, ?> windowingStrategy,
+                               final Map<Integer, PCollectionView<?>> sideInputs,
+                               final PipelineOptions options,
+                               final DisplayData displayData) {
+    super(doFn, inputCoder, outputCoders, mainOutputTag,
+      additionalOutputTags, windowingStrategy, sideInputs, options, displayData);
+    this.curPushedBacks = new ArrayList<>();
+    this.curPushedBackWatermark = Long.MAX_VALUE;
+    this.curInputWatermark = Long.MIN_VALUE;
+    this.curOutputWatermark = Long.MIN_VALUE;
+  }
+
+  @Override
+  protected DoFn wrapDoFn(final DoFn initDoFn) {
+    return initDoFn;
+  }
+
+  @Override
+  public void onData(final WindowedValue data) {
+    // Need to distinguish side/main inputs and push-back main inputs.
+    if (data.getValue() instanceof SideInputElement) {
+      // This element is a Side Input
+      // TODO #287: Consider Explicit Multi-Input IR Transform
+      final WindowedValue<SideInputElement> sideInputElement = (WindowedValue<SideInputElement>) data;
+      final PCollectionView view = getSideInputs().get(sideInputElement.getValue().getSideInputIndex());
+      getSideInputReader().addSideInputElement(view, data);
+
+      handlePushBacks();
+
+      // See if we can emit a new watermark, as we may have processed some pushed-back elements
+      onWatermark(new Watermark(curInputWatermark));
+    } else {
+      // This element is the Main Input
+      checkAndInvokeBundle();
+      final Iterable<WindowedValue<InputT>> pushedBack =
+        getPushBackRunner().processElementInReadyWindows(data);
+      for (final WindowedValue wv : pushedBack) {
+        curPushedBackWatermark = Math.min(curPushedBackWatermark, wv.getTimestamp().getMillis());
+        curPushedBacks.add(wv);
+      }
+      checkAndFinishBundle();
+    }
+  }
+
+  private void handlePushBacks() {
+    // Force-flush, before (possibly) processing pushed-back data.
+    //
+    // Main reason:
+    // {@link org.apache.beam.runners.core.SimplePushbackSideInputDoFnRunner}
+    // caches for each bundle the side inputs that are not ready.
+    // We need to re-start the bundle to advertise the (possibly) newly available side input.
+    forceFinishBundle(); // forced
+
+    checkAndInvokeBundle();
+    // With the new side input added, we may be able to process some pushed-back elements.
+    final List<WindowedValue<InputT>> pushedBackAgain = new ArrayList<>();
+    long pushedBackAgainWatermark = Long.MAX_VALUE;
```

Comment:
curPushBackWatermark = Long.MAX_VALUE

Reply:
Yes, but  I think we don't need the `pushedBackAgainWatermark` variable if we just set `curPushBackWatermark = Long.MAX_VALUE` here?